File: js/test/audit_queue.test.mjs
/**
 * P1-15 cross-language: same wire/API semantics as Python's
 * tests/test_audit_queue.py — happy path, slow store, outage+recovery,
 * raise mode, sys self-report, queue_stats, flush.
 *
 * JS deviation: emit() doesn't block on capacity (single-threaded
 * event loop). queueCapacity is the high-water-warn threshold; the
 * drainer + retry-forever-with-backoff matches Python exactly.
 */
import { test, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import {
    init, emit, register, registry as _reg,
    AuditStoreError, queueStats, flush,
} from '../src/index.js';
import { uninstallDrainer } from '../src/audit_queue.js';

// ── stub stores ────────────────────────────────────────────────────────────

class RecordingStore {
    constructor() { this.rows = []; }
    insert(row) { this.rows.push(row); }
}

class SlowStore extends RecordingStore {
    constructor(delayMs) { super(); this.delayMs = delayMs; }
    async insert(row) {
        await new Promise(r => setTimeout(r, this.delayMs));
        super.insert(row);
    }
}

class BrokenStore {
    constructor(msg = 'store down') { this.msg = msg; this.attempts = 0; }
    insert(_row) {
        this.attempts++;
        throw new Error(this.msg);
    }
}

class FlakyStore {
    constructor(failFirst) {
        this.failFirst = failFirst;
        this.attempts = 0;
        this.success = 0;
    }
    insert(_row) {
        this.attempts++;
        if (this.attempts <= this.failFirst) {
            throw new Error(`transient #${this.attempts}`);
        }
        this.success++;
    }
}

// ── fixtures ───────────────────────────────────────────────────────────────

const TEST_META = {
    USER_CREATED: {
        id: 'USER_CREATED', domain: 'user', category: 'account',
        action: 'create', severity: 'info',
        retention_class: 'short', description: 'test', emitter: 'test',
    },
};

beforeEach(() => {
    uninstallDrainer();
    _reg().clear();
    register('user', TEST_META);
});

afterEach(() => {
    uninstallDrainer();
});

function _initQueue(store, opts = {}) {
    init({
        serviceId: 'svc',
        nodeId: 'node',
        auditStore: store,
        auditStoreFailureStrategy: 'queue',
        queueCapacity: 100,
        ...opts,
    });
}

function _initRaise(store) {
    init({
        serviceId: 'svc',
        nodeId: 'node',
        auditStore: store,
        auditStoreFailureStrategy: 'raise',
    });
}

// ── #1 happy path ──────────────────────────────────────────────────────────

test('queue mode drains within one second', async () => {
    const store = new RecordingStore();
    _initQueue(store);
    emit({ code: 'USER_CREATED', target: 'u-1' });
    emit({ code: 'USER_CREATED', target: 'u-2' });
    assert.equal(await flush(1000), true, 'flush should succeed');
    assert.equal(store.rows.length, 2);
});

test('queue mode emit returns immediately under slow store', async () => {
    const store = new SlowStore(50);
    _initQueue(store);
    const t0 = Date.now();
    for (let i = 0; i < 10; i++) emit({ code: 'USER_CREATED', target: `u-${i}` });
    const elapsed = Date.now() - t0;
    assert.ok(elapsed < 100, `emit() should return immediately; took ${elapsed}ms`);
    assert.equal(await flush(2000), true, 'flush should drain the slow store');
    assert.equal(store.rows.length, 10);
});

// ── #2 raise mode ──────────────────────────────────────────────────────────

test('raise mode throws AuditStoreError on failure', () => {
    const store = new BrokenStore();
    _initRaise(store);
    assert.throws(
        () => emit({ code: 'USER_CREATED', target: 'u-1' }),
        (err) => err instanceof AuditStoreError && /store down/.test(err.message),
    );
});

test('raise mode does not install drainer', () => {
    const store = new RecordingStore();
    _initRaise(store);
    assert.equal(queueStats(), null);
});

// ── #3 outage + recovery ───────────────────────────────────────────────────

test('outage then recovery drains pending rows', async () => {
    const store = new FlakyStore(2);
    _initQueue(store, {
        queueRetryInitialMs: 10,
        queueRetryMaxMs: 50,
        queueRetryJitter: false,
    });
    emit({ code: 'USER_CREATED', target: 'u-1' });
    assert.equal(await flush(2000), true, 'drainer should retry past 2 failures');
    assert.equal(store.success, 1);
    assert.ok(store.attempts >= 3, `want >=3 attempts (2 fail + 1 success); got ${store.attempts}`);
});

// ── #5 sys self-report ────────────────────────────────────────────────────

test('drainer first failure emits warn to sys (captured via stdout)', async () => {
    const store = new FlakyStore(1);
    const captured = [];
    const origWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = (chunk, ...rest) => {
        try {
            const obj = JSON.parse(String(chunk).trim());
            if (obj.shape === 'sys') captured.push(obj);
        } catch { /* not json */ }
        return origWrite(chunk, ...rest);
    };
    try {
        _initQueue(store, {
            queueRetryInitialMs: 5,
            queueRetryMaxMs: 20,
            queueRetryJitter: false,
        });
        emit({ code: 'USER_CREATED', target: 'u-1' });
        assert.equal(await flush(2000), true);
    } finally {
        process.stdout.write = origWrite;
    }
    const events = captured.map(r => r.event);
    assert.ok(events.includes('audit_drain_failed'),
        `expected audit_drain_failed; got ${JSON.stringify(events)}`);
    assert.ok(events.includes('audit_drain_recovered'),
        `expected audit_drain_recovered; got ${JSON.stringify(events)}`);
});

test('drainer degraded after 5 failures', async () => {
    const store = new BrokenStore();
    const captured = [];
    const origWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = (chunk, ...rest) => {
        try {
            const obj = JSON.parse(String(chunk).trim());
            if (obj.shape === 'sys') captured.push(obj);
        } catch { /* not json */ }
        return origWrite(chunk, ...rest);
    };
    try {
        _initQueue(store, {
            queueRetryInitialMs: 5,
            queueRetryMaxMs: 20,
            queueRetryJitter: false,
        });
        emit({ code: 'USER_CREATED', target: 'u-1' });
        const deadline = Date.now() + 3000;
        while (Date.now() < deadline) {
            if (captured.some(r => r.event === 'audit_drain_degraded')) break;
            await new Promise(r => setTimeout(r, 50));
        }
    } finally {
        process.stdout.write = origWrite;
    }
    assert.ok(
        captured.some(r => r.event === 'audit_drain_degraded'),
        `audit_drain_degraded never fired; got ${JSON.stringify(captured.map(c => c.event))}`,
    );
});

test('queue high water emits warn at 50%', async () => {
    const store = new BrokenStore();
    const captured = [];
    const origWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = (chunk, ...rest) => {
        try {
            const obj = JSON.parse(String(chunk).trim());
            if (obj.shape === 'sys') captured.push(obj);
        } catch { /* not json */ }
        return origWrite(chunk, ...rest);
    };
    try {
        _initQueue(store, {
            queueCapacity: 4,
            queueRetryInitialMs: 5_000,
            queueRetryJitter: false,
        });
        for (let i = 0; i < 2; i++) emit({ code: 'USER_CREATED', target: `u-${i}` });
        await new Promise(r => setTimeout(r, 100));
    } finally {
        process.stdout.write = origWrite;
    }
    assert.ok(
        captured.some(r => r.event === 'audit_queue_high_water'),
        `expected audit_queue_high_water; got ${JSON.stringify(captured.map(c => c.event))}`,
    );
});

// ── #6 queueStats + flush ──────────────────────────────────────────────────

test('public flush blocks until drained', async () => {
    const store = new RecordingStore();
    _initQueue(store);
    for (let i = 0; i < 5; i++) emit({ code: 'USER_CREATED', target: `u-${i}` });
    assert.equal(await flush(2000), true);
    assert.equal(store.rows.length, 5);
});

test('public flush is no-op in raise mode', async () => {
    const store = new RecordingStore();
    _initRaise(store);
    emit({ code: 'USER_CREATED', target: 'u-1' });
    assert.equal(await flush(100), true);
});

test('queue stats fields', async () => {
    const store = new RecordingStore();
    _initQueue(store);
    emit({ code: 'USER_CREATED', target: 'u-1' });
    assert.equal(await flush(1000), true);
    const s = queueStats();
    assert.ok(s !== null, 'expected non-null stats');
    assert.equal(s.depth, 0);
    assert.equal(s.capacity, 100);
    assert.ok(s.high_water >= 1);
    assert.equal(s.drained_total, 1);
    assert.equal(s.retry_count_active, 0);
    assert.equal(s.last_error, null);
});

test('queue stats high water monotonic', async () => {
    const store = new RecordingStore();
    _initQueue(store, { queueCapacity: 10 });
    for (let i = 0; i < 5; i++) emit({ code: 'USER_CREATED', target: `u-${i}` });
    assert.equal(await flush(1000), true);
    const high = queueStats().high_water;
    emit({ code: 'USER_CREATED', target: 'u-x' });
    assert.equal(await flush(1000), true);
    assert.ok(queueStats().high_water >= high);
});

File: go/audit_queue_test.go
// P1-15 cross-language: same wire/API semantics as Python's
// tests/test_audit_queue.py — happy path, slow store, outage+recovery,
// queue full blocks, raise mode, sys self-report, queue_stats, flush.
package fasten

import (
	"context"
	"errors"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// ── Stub stores ────────────────────────────────────────────────────────────

type recordingStore struct {
	mu   sync.Mutex
	rows []Row
}

func (s *recordingStore) Insert(_ context.Context, r Row) error {
	s.mu.Lock()
	s.rows = append(s.rows, r)
	s.mu.Unlock()
	return nil
}
func (s *recordingStore) Query(_ context.Context, _ Filter) ([]Row, error) { return nil, nil }
func (s *recordingStore) ListUnshipped(_ context.Context, _ int) ([]Row, error) {
	return nil, nil
}
func (s *recordingStore) MarkShipped(_ context.Context, _ []string) error    { return nil }
func (s *recordingStore) Purge(_ context.Context, _ time.Time, _ bool) (int, error) { return 0, nil }

func (s *recordingStore) count() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.rows)
}

type slowStore struct {
	recordingStore
	delay time.Duration
}

func (s *slowStore) Insert(ctx context.Context, r Row) error {
	time.Sleep(s.delay)
	return s.recordingStore.Insert(ctx, r)
}

type brokenStore struct {
	mu       sync.Mutex
	attempts int32
	msg      string
}

func (s *brokenStore) Insert(_ context.Context, _ Row) error {
	atomic.AddInt32(&s.attempts, 1)
	return errors.New(s.msg)
}
func (s *brokenStore) Query(_ context.Context, _ Filter) ([]Row, error) { return nil, nil }
func (s *brokenStore) ListUnshipped(_ context.Context, _ int) ([]Row, error) {
	return nil, nil
}
func (s *brokenStore) MarkShipped(_ context.Context, _ []string) error    { return nil }
func (s *brokenStore) Purge(_ context.Context, _ time.Time, _ bool) (int, error) { return 0, nil }

type flakyStore struct {
	mu        sync.Mutex
	attempts  int32
	success   int32
	failFirst int32
}

func (s *flakyStore) Insert(_ context.Context, _ Row) error {
	n := atomic.AddInt32(&s.attempts, 1)
	if n <= s.failFirst {
		return errors.New("transient")
	}
	atomic.AddInt32(&s.success, 1)
	return nil
}
func (s *flakyStore) Query(_ context.Context, _ Filter) ([]Row, error) { return nil, nil }
func (s *flakyStore) ListUnshipped(_ context.Context, _ int) ([]Row, error) {
	return nil, nil
}
func (s *flakyStore) MarkShipped(_ context.Context, _ []string) error    { return nil }
func (s *flakyStore) Purge(_ context.Context, _ time.Time, _ bool) (int, error) { return 0, nil }

// ── Test helpers ───────────────────────────────────────────────────────────

func resetState() {
	Default.ResetForTests()
	regMu.Lock()
	_registry = map[Code]Meta{}
	regMu.Unlock()
}

func setupQueueMode(t *testing.T, store AuditRepository, opts ...func(*Config)) {
	t.Helper()
	resetState()
	if err := Register("user", map[Code]Meta{
		"USER_CREATED": {
			Domain:         "user",
			Category:       "account",
			Action:         "create",
			Severity:       SevInfo,
			Description:    "test",
			Emitter:        "test",
			RetentionClass: RetLong,
		},
	}); err != nil {
		t.Fatalf("register: %v", err)
	}
	cfg := Config{
		ServiceID:                 "svc",
		NodeID:                    "node",
		AuditStore:                store,
		AuditStoreFailureStrategy: "queue",
		QueueCapacity:             100,
	}
	for _, o := range opts {
		o(&cfg)
	}
	if err := Init(cfg); err != nil {
		t.Fatalf("init: %v", err)
	}
	t.Cleanup(resetState)
}

func setupRaiseMode(t *testing.T, store AuditRepository) {
	t.Helper()
	resetState()
	if err := Register("user", map[Code]Meta{
		"USER_CREATED": {
			Domain: "user", Category: "account", Action: "create",
			Severity: SevInfo, Description: "test", Emitter: "test",
			RetentionClass: RetLong,
		},
	}); err != nil {
		t.Fatalf("register: %v", err)
	}
	if err := Init(Config{
		ServiceID:                 "svc",
		NodeID:                    "node",
		AuditStore:                store,
		AuditStoreFailureStrategy: "raise",
	}); err != nil {
		t.Fatalf("init: %v", err)
	}
	t.Cleanup(resetState)
}

// ── #1 happy path ──────────────────────────────────────────────────────────

func TestQueueModeDrainsWithinOneSecond(t *testing.T) {
	store := &recordingStore{}
	setupQueueMode(t, store)
	ctx := context.Background()
	if _, err := Emit(ctx, "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if _, err := Emit(ctx, "USER_CREATED", Target("u-2")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(time.Second) {
		t.Fatal("flush timed out")
	}
	if store.count() != 2 {
		t.Fatalf("want 2 rows; got %d", store.count())
	}
}

func TestQueueModeEmitReturnsImmediatelyUnderSlowStore(t *testing.T) {
	store := &slowStore{delay: 50 * time.Millisecond}
	setupQueueMode(t, store)
	ctx := context.Background()
	t0 := time.Now()
	for i := 0; i < 10; i++ {
		if _, err := Emit(ctx, "USER_CREATED", Target("u")); err != nil {
			t.Fatalf("emit: %v", err)
		}
	}
	elapsed := time.Since(t0)
	// 10 emits * 50ms sync would be 500ms+. Async should be < 100ms.
	if elapsed >= 100*time.Millisecond {
		t.Fatalf("emit() should return immediately; took %s", elapsed)
	}
	if !Flush(2 * time.Second) {
		t.Fatal("flush timed out")
	}
	if store.count() != 10 {
		t.Fatalf("want 10 rows; got %d", store.count())
	}
}

// ── #2 raise mode ──────────────────────────────────────────────────────────

func TestRaiseModeReturnsAuditStoreError(t *testing.T) {
	store := &brokenStore{msg: "store down"}
	setupRaiseMode(t, store)
	_, err := Emit(context.Background(), "USER_CREATED", Target("u-1"))
	if err == nil {
		t.Fatal("expected error in raise mode")
	}
	var aerr *AuditStoreError
	if !errors.As(err, &aerr) {
		t.Fatalf("expected *AuditStoreError; got %T", err)
	}
	if aerr.Err.Error() != "store down" {
		t.Fatalf("unwrapped: %v", aerr.Err)
	}
}

func TestRaiseModeDoesNotInstallDrainer(t *testing.T) {
	store := &recordingStore{}
	setupRaiseMode(t, store)
	if Default.activeDrainer() != nil {
		t.Fatal("raise mode must not install a drainer")
	}
	if GetQueueStats() != nil {
		t.Fatal("queue_stats must be nil in raise mode")
	}
}

// ── #3 outage + recovery ───────────────────────────────────────────────────

func TestOutageThenRecoveryDrainsPendingRows(t *testing.T) {
	store := &flakyStore{failFirst: 2}
	setupQueueMode(t, store, func(c *Config) {
		c.QueueRetryInitial = 10 * time.Millisecond
		c.QueueRetryMax = 50 * time.Millisecond
		c.DisableQueueJitter = true
	})
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(2 * time.Second) {
		t.Fatal("flush timed out — drainer did not recover")
	}
	if atomic.LoadInt32(&store.success) != 1 {
		t.Fatalf("want 1 success; got %d", store.success)
	}
	if atomic.LoadInt32(&store.attempts) < 3 {
		t.Fatalf("want >=3 attempts (2 fail + 1 success); got %d", store.attempts)
	}
}

// ── #4 queue full blocks emit ──────────────────────────────────────────────

func TestQueueFullBlocksEmitDoesNotSilentlyDrop(t *testing.T) {
	store := &brokenStore{msg: "store down"}
	setupQueueMode(t, store, func(c *Config) {
		c.QueueCapacity = 2
		c.QueueRetryInitial = 10 * time.Second // never retries during the test
		c.DisableQueueJitter = true
	})
	ctx := context.Background()
	if _, err := Emit(ctx, "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit-1: %v", err)
	}
	if _, err := Emit(ctx, "USER_CREATED", Target("u-2")); err != nil {
		t.Fatalf("emit-2: %v", err)
	}

	finished := make(chan struct{})
	blocked := make(chan struct{})
	go func() {
		close(blocked)
		_, _ = Emit(ctx, "USER_CREATED", Target("u-3"))
		close(finished)
	}()
	<-blocked
	select {
	case <-finished:
		t.Fatal("emit() must block when queue is full and store is unavailable")
	case <-time.After(200 * time.Millisecond):
		// good — still blocked
	}

	// Stopping the drainer pops rows + releases slots → unblocks emit.
	Default.uninstallDrainer()
	select {
	case <-finished:
	case <-time.After(2 * time.Second):
		t.Fatal("emit should unblock once queue space frees")
	}
}

// ── #5 sys self-report ────────────────────────────────────────────────────

func TestDrainerFirstFailureEmitsWarnToSys(t *testing.T) {
	store := &flakyStore{failFirst: 1}
	setupQueueMode(t, store, func(c *Config) {
		c.QueueRetryInitial = 5 * time.Millisecond
		c.QueueRetryMax = 20 * time.Millisecond
		c.DisableQueueJitter = true
	})
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(2 * time.Second) {
		t.Fatal("flush timed out")
	}
	rows := Default.xport.QuerySyslog(100, "", "", "")
	events := map[string]bool{}
	for _, r := range rows {
		if e, ok := r["event"].(string); ok {
			events[e] = true
		}
	}
	if !events["audit_drain_failed"] {
		t.Fatalf("expected first-failure warn; got events %v", events)
	}
	if !events["audit_drain_recovered"] {
		t.Fatalf("expected recovery info; got events %v", events)
	}
}

func TestDrainerDegradedAfter5Failures(t *testing.T) {
	store := &brokenStore{msg: "store down"}
	setupQueueMode(t, store, func(c *Config) {
		c.QueueRetryInitial = 5 * time.Millisecond
		c.QueueRetryMax = 20 * time.Millisecond
		c.DisableQueueJitter = true
	})
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	deadline := time.Now().Add(3 * time.Second)
	for time.Now().Before(deadline) {
		rows := Default.xport.QuerySyslog(100, "", "", "")
		for _, r := range rows {
			if r["event"] == "audit_drain_degraded" {
				return
			}
		}
		time.Sleep(50 * time.Millisecond)
	}
	t.Fatal("audit_drain_degraded never fired after 5+ failures")
}

func TestQueueHighWaterEmitsWarnAt50pct(t *testing.T) {
	store := &brokenStore{msg: "store down"}
	setupQueueMode(t, store, func(c *Config) {
		c.QueueCapacity = 4
		c.QueueRetryInitial = 5 * time.Second
		c.DisableQueueJitter = true
	})
	for i := 0; i < 2; i++ {
		if _, err := Emit(context.Background(), "USER_CREATED", Target("u")); err != nil {
			t.Fatalf("emit: %v", err)
		}
	}
	time.Sleep(100 * time.Millisecond)
	rows := Default.xport.QuerySyslog(100, "", "", "")
	events := map[string]bool{}
	for _, r := range rows {
		if e, ok := r["event"].(string); ok {
			events[e] = true
		}
	}
	if !events["audit_queue_high_water"] {
		t.Fatalf("expected high-water warn; got events %v", events)
	}
}

// ── #6 queue_stats + flush ─────────────────────────────────────────────────

func TestPublicFlushBlocksUntilDrained(t *testing.T) {
	store := &recordingStore{}
	setupQueueMode(t, store)
	for i := 0; i < 5; i++ {
		if _, err := Emit(context.Background(), "USER_CREATED", Target("u")); err != nil {
			t.Fatalf("emit: %v", err)
		}
	}
	if !Flush(2 * time.Second) {
		t.Fatal("flush timed out")
	}
	if store.count() != 5 {
		t.Fatalf("want 5; got %d", store.count())
	}
}

func TestPublicFlushNoOpInRaiseMode(t *testing.T) {
	store := &recordingStore{}
	setupRaiseMode(t, store)
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(100 * time.Millisecond) {
		t.Fatal("Flush() must return true in raise mode (no queue)")
	}
}

func TestQueueStatsFields(t *testing.T) {
	store := &recordingStore{}
	setupQueueMode(t, store)
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(time.Second) {
		t.Fatal("flush timed out")
	}
	s := GetQueueStats()
	if s == nil {
		t.Fatal("expected non-nil stats")
	}
	if s.Depth != 0 {
		t.Errorf("depth want 0; got %d", s.Depth)
	}
	if s.Capacity != 100 {
		t.Errorf("capacity want 100; got %d", s.Capacity)
	}
	if s.HighWater < 1 {
		t.Errorf("high_water want >=1; got %d", s.HighWater)
	}
	if s.DrainedTotal != 1 {
		t.Errorf("drained_total want 1; got %d", s.DrainedTotal)
	}
	if s.RetryCountActive != 0 {
		t.Errorf("retry_count_active want 0; got %d", s.RetryCountActive)
	}
	if s.LastError != "" {
		t.Errorf("last_error want empty; got %q", s.LastError)
	}
}

func TestQueueStatsHighWaterMonotonic(t *testing.T) {
	store := &recordingStore{}
	setupQueueMode(t, store, func(c *Config) { c.QueueCapacity = 10 })
	for i := 0; i < 5; i++ {
		if _, err := Emit(context.Background(), "USER_CREATED", Target("u")); err != nil {
			t.Fatalf("emit: %v", err)
		}
	}
	if !Flush(time.Second) {
		t.Fatal("flush timed out")
	}
	high := GetQueueStats().HighWater
	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-x")); err != nil {
		t.Fatalf("emit: %v", err)
	}
	if !Flush(time.Second) {
		t.Fatal("flush timed out")
	}
	if got := GetQueueStats().HighWater; got < high {
		t.Errorf("high_water regressed: %d → %d", high, got)
	}
}

// REVIEW #19: AuditStoreFailureStrategy must accept any case (parity
// with Python). Earlier the switch matched only lowercase, so "Queue"
// or "RAISE" silently failed init.
func TestInit_StrategyCaseInsensitive(t *testing.T) {
	for _, s := range []string{"Queue", "QUEUE", "Raise", "RAISE", "qUeUe"} {
		t.Run(s, func(t *testing.T) {
			resetState()
			if err := Register("user", map[Code]Meta{
				"USER_CREATED": {
					Domain: "user", Category: "account", Action: "create",
					Severity: SevInfo, Description: "test", Emitter: "test",
					RetentionClass: RetLong,
				},
			}); err != nil {
				t.Fatalf("register: %v", err)
			}
			err := Init(Config{
				ServiceID:                 "svc",
				NodeID:                    "node",
				AuditStore:                &recordingStore{},
				AuditStoreFailureStrategy: s,
			})
			if err != nil {
				t.Fatalf("strategy %q rejected: %v", s, err)
			}
			t.Cleanup(resetState)
		})
	}
}

func TestInit_StrategyInvalidValueErrorMentionsValue(t *testing.T) {
	resetState()
	defer resetState()
	if err := Register("user", map[Code]Meta{
		"USER_CREATED": {
			Domain: "user", Category: "account", Action: "create",
			Severity: SevInfo, Description: "test", Emitter: "test",
			RetentionClass: RetLong,
		},
	}); err != nil {
		t.Fatalf("register: %v", err)
	}
	err := Init(Config{
		ServiceID:                 "svc",
		NodeID:                    "node",
		AuditStore:                &recordingStore{},
		AuditStoreFailureStrategy: "weird",
	})
	if err == nil {
		t.Fatal("init must reject unknown strategy")
	}
	// Error must include the offending value so adopters can debug
	// typos without re-reading the docs.
	msg := err.Error()
	if !contains(msg, "\"weird\"") {
		t.Errorf("error must mention the bad value; got: %q", msg)
	}
}

// REVIEW #20: ring buffer used `rb.buf = rb.buf[1:]` to drop the
// oldest entry, which only advanced the slice header — the original
// backing array kept growing one element per push past capacity.
// Now uses a true circular index. Push past 10x capacity must keep
// the underlying array length bounded at exactly cap.
func TestRingBuffer_NoUnboundedGrowth(t *testing.T) {
	rb := newRingBuffer[int](16)
	for i := 0; i < 16*100; i++ {
		rb.Push(i)
	}
	if got := rb.Len(); got != 16 {
		t.Errorf("logical Len() want 16; got %d", got)
	}
	if got := cap(rb.buf); got != 16 {
		t.Errorf("backing array cap should stay at the configured size 16; got %d", got)
	}
	// Snapshot is newest-first and contains exactly 16 entries.
	all := rb.All()
	if len(all) != 16 {
		t.Fatalf("All() len want 16; got %d", len(all))
	}
	if all[0] != 16*100-1 {
		t.Errorf("newest entry want %d; got %d", 16*100-1, all[0])
	}
	if all[15] != 16*100-16 {
		t.Errorf("oldest of the 16 retained want %d; got %d", 16*100-16, all[15])
	}
}

// ── P1-22: Go drainer panic-recovery ─────────────────────────────────────

type panicStore struct{}

func (s *panicStore) Insert(_ context.Context, _ Row) error  { panic("store exploded") }
func (s *panicStore) Query(_ context.Context, _ Filter) ([]Row, error) { return nil, nil }
func (s *panicStore) ListUnshipped(_ context.Context, _ int) ([]Row, error) {
	return nil, nil
}
func (s *panicStore) MarkShipped(_ context.Context, _ []string) error    { return nil }
func (s *panicStore) Purge(_ context.Context, _ time.Time, _ bool) (int, error) { return 0, nil }

// TestDrainerSurvivesStorePanic asserts that a panicking store.Insert emits
// audit_drain_recovered_from_panic to the sys stream and that subsequent
// Emit + QueueStats calls do not panic or hang.
func TestDrainerSurvivesStorePanic(t *testing.T) {
	setupQueueMode(t, &panicStore{}, func(c *Config) {
		c.QueueRetryInitial = 5 * time.Millisecond
		c.QueueRetryMax = 20 * time.Millisecond
		c.DisableQueueJitter = true
	})

	if _, err := Emit(context.Background(), "USER_CREATED", Target("u-1")); err != nil {
		t.Fatalf("emit: %v", err)
	}

	// Allow time for the drainer goroutine to run and recover.
	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		rows := Default.xport.QuerySyslog(100, "", "", "")
		for _, r := range rows {
			if r["event"] == "audit_drain_recovered_from_panic" {
				// Drainer survived. Verify subsequent calls don't hang or panic.
				_, _ = Emit(context.Background(), "USER_CREATED", Target("u-2"))
				_ = GetQueueStats()
				return
			}
		}
		time.Sleep(20 * time.Millisecond)
	}
	t.Fatal("audit_drain_recovered_from_panic sys event never fired after store panic")
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
